File: src/lib/workout-streak.ts
import dayjs from "dayjs";
import utc from "dayjs/plugin/utc.js";

import { WeekDay } from "../generated/prisma/enums.js";

dayjs.extend(utc);

const WEEKDAY_BY_DAY_INDEX: Record<number, WeekDay> = {
  0: WeekDay.SUNDAY,
  1: WeekDay.MONDAY,
  2: WeekDay.TUESDAY,
  3: WeekDay.WEDNESDAY,
  4: WeekDay.THURSDAY,
  5: WeekDay.FRIDAY,
  6: WeekDay.SATURDAY,
};

const MAX_STREAK_LOOKBACK_IN_DAYS = 365;

interface CalculateWorkoutStreakParams {
  workoutDays: Array<{ weekDay: WeekDay; isRest: boolean }>;
  completedDates: Set<string>; // YYYY-MM-DD (UTC)
  referenceDate: dayjs.Dayjs;
  planCreatedAt: Date;
}

export const getWeekDay = (date: dayjs.Dayjs): WeekDay =>
  WEEKDAY_BY_DAY_INDEX[date.day()];

export const getCompletedSessionDates = (
  workoutDays: Array<{
    sessions: Array<{ startedAt: Date; completedAt: Date | null }>;
  }>
): Set<string> =>
  new Set(
    workoutDays.flatMap((day) =>
      day.sessions
        .filter((session) => session.completedAt !== null)
        .map((session) => dayjs.utc(session.startedAt).format("YYYY-MM-DD"))
    )
  );

export const calculateWorkoutStreak = ({
  workoutDays,
  completedDates,
  referenceDate,
  planCreatedAt,
}: CalculateWorkoutStreakParams): number => {
  const planWeekDays = new Set(workoutDays.map((day) => day.weekDay));
  const restWeekDays = new Set(
    workoutDays.filter((day) => day.isRest).map((day) => day.weekDay)
  );
  const reference = referenceDate.utc().startOf("day");
  const planStart = dayjs.utc(planCreatedAt).startOf("day");

  let streak = 0;

  for (let offset = 0; offset < MAX_STREAK_LOOKBACK_IN_DAYS; offset++) {
    const day = reference.subtract(offset, "day");
    if (day.isBefore(planStart)) {
      break;
    }

    const weekDay = getWeekDay(day);
    if (!planWeekDays.has(weekDay)) {
      continue;
    }

    const isCompleted = completedDates.has(day.format("YYYY-MM-DD"));
    if (restWeekDays.has(weekDay) || isCompleted) {
      streak++;
      continue;
    }

    // O dia de referência (hoje) ainda pode ser concluído: não quebra a sequência.
    if (offset === 0) {
      continue;
    }

    break;
  }

  return streak;
};

File: backend/FitAi.Api/Domain/WorkoutStreak.cs
using FitAi.Contracts;

namespace FitAi.Api.Domain;

/// <summary>Cálculo da sequência (🔥) — função pura, sem acesso ao banco. Datas em UTC.</summary>
public static class WorkoutStreak
{
    private const int MaxLookbackInDays = 365;

    public static WeekDay GetWeekDay(DateOnly date) => date.DayOfWeek switch
    {
        DayOfWeek.Sunday => WeekDay.SUNDAY,
        DayOfWeek.Monday => WeekDay.MONDAY,
        DayOfWeek.Tuesday => WeekDay.TUESDAY,
        DayOfWeek.Wednesday => WeekDay.WEDNESDAY,
        DayOfWeek.Thursday => WeekDay.THURSDAY,
        DayOfWeek.Friday => WeekDay.FRIDAY,
        _ => WeekDay.SATURDAY,
    };

    public static DateOnly ToUtcDate(DateTimeOffset value) => DateOnly.FromDateTime(value.UtcDateTime);

    public static string ToKey(DateOnly date) => date.ToString("yyyy-MM-dd");

    public static HashSet<DateOnly> GetCompletedDates(IEnumerable<(DateTimeOffset StartedAt, DateTimeOffset? CompletedAt)> sessions) =>
        sessions.Where(s => s.CompletedAt is not null).Select(s => ToUtcDate(s.StartedAt)).ToHashSet();

    /// <summary>
    /// Conta, da data de referência para trás, os dias seguidos cumpridos. Dia de descanso conta;
    /// dia fora do plano é pulado; o dia de referência (hoje) não quebra a sequência.
    /// </summary>
    public static int Calculate(
        IEnumerable<(WeekDay WeekDay, bool IsRest)> workoutDays,
        IReadOnlySet<DateOnly> completedDates,
        DateOnly referenceDate,
        DateOnly planCreatedAt)
    {
        var days = workoutDays.ToList();
        var planWeekDays = days.Select(d => d.WeekDay).ToHashSet();
        var restWeekDays = days.Where(d => d.IsRest).Select(d => d.WeekDay).ToHashSet();

        var streak = 0;
        for (var offset = 0; offset < MaxLookbackInDays; offset++)
        {
            var day = referenceDate.AddDays(-offset);
            if (day < planCreatedAt) break;

            var weekDay = GetWeekDay(day);
            if (!planWeekDays.Contains(weekDay)) continue;

            if (restWeekDays.Contains(weekDay) || completedDates.Contains(day))
            {
                streak++;
                continue;
            }

            // O dia de referência (hoje) ainda pode ser concluído: não quebra a sequência.
            if (offset == 0) continue;

            break;
        }

        return streak;
    }
}
